i6_models/primitives/feature_extraction.py

```diff
@@ -111,3 +117,107 @@ def forward(self, raw_audio, length) -> Tuple[torch.Tensor, torch.Tensor]:
+        super().__post_init__()
+        assert self.win_size > 0 and self.hop_size > 0, "window settings need to be positive"
+        assert self.num_filters > 0, "number of filters needs to be positive"
+        assert self.hop_size <= self.win_size, "using a larger hop size than window size does not make sense"
+
+
+class RasrCompatibleLogMelFeatureExtractionV1(nn.Module):
+    """
+    Rasr-compatible log-mel feature extraction using log10. Does not use torchaudio.
+    """
+
+    def __init__(self, cfg: RasrCompatibleLogMelFeatureExtractionV1Config):
+        super().__init__()
+
+        self.sample_rate = int(cfg.sample_rate)
+        self.hop_length = int(cfg.hop_size * cfg.sample_rate)
+        self.min_amp = cfg.min_amp
+        self.win_length = int(cfg.win_size * cfg.sample_rate)
+        self.n_fft = 2 ** math.ceil(
+            math.log2(self.win_length)
+        )  # smallest power if two which is greater than or equal to win_length
+        self.alpha = cfg.alpha
+
+        self.register_buffer(
+            "mel_basis",
+            torch.tensor(
+                filters.mel(
+                    sr=cfg.sample_rate,
+                    n_fft=self.n_fft,
+                    n_mels=cfg.num_filters,
+                    fmin=0,
+                    fmax=cfg.sample_rate // 2,
+                    htk=True,
+                    norm=None,
+                ),
+            ),
+        )
+        self.register_buffer(
+            "window", torch.hann_window(self.win_length, periodic=False, dtype=torch.float64).to(torch.float32)
+        )
+
+    def forward(self, raw_audio, length) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        :param raw_audio: [B, T]
+        :param length: in samples [B]
+        :return features as [B,T,F] and length in frames [B]
+        """
+        assert raw_audio.shape[1] > 0  # also same for length
+        res_size = max(raw_audio.shape[1] - self.win_length + self.hop_length - 1, 0) // self.hop_length + 1
+        res_length = (
+            torch.maximum(length - self.win_length + self.hop_length - 1, torch.zeros_like(length)) // self.hop_length
+            + 1
+        )
+
+        # preemphasize
+        preemphasized = raw_audio.clone()
+        preemphasized[..., 1:] -= self.alpha * preemphasized[..., :-1]
+        preemphasized[..., 0] = 0.0
+
+        # zero pad for the last frame
+        last_win_size = preemphasized.shape[1] - (res_size - 1) * self.hop_length
+        last_pad = self.win_length - last_win_size
+        padded = torch.nn.functional.pad(preemphasized, (0, last_pad))
+
+        windowed = padded.unfold(1, size=self.win_length, step=self.hop_length)  # [B, T', W=win_length]
+        smoothed = windowed[:, :-1] * self.window[None, None, :]  # [B, T'-1, W]
+
+        # The last window might be shorter. Will use a shorter Hanning window then. Need to fix that.
+        last_win = torch.hann_window(last_win_size, periodic=False, dtype=torch.float64).to(
```

Comment:
Is this correct if you have sequences of different lengths in the batch? Maybe it's necessary to do this as a loop over all sequences in the batch and modify some parts of the output with this?